### PR TITLE
Warn on transitive macro recompilation, with opt-out

### LINKED ISSUE #142: Forward port Silence macro recompilation info message
https://github.com/sbt/sbt/pull/2659

### PR DESCRIPTION
**Pull request by dwijnand**
Forward port of sbt/sbt#2637 and sbt/sbt#2659. Fixes #142

A forward port of the test (hopefully) to follow.

---

I would love to have a test for this, like I do for sbt 0.13, but I would need some help, as I don't know how things are done in this codebase. Thanks.

### CODE REVIEW

File: zinc/src/sbt-test/reporter/warnings-reported/test [hunk 1]
```diff
@@ -0,0 +1,4 @@
+> compile
+$ pause
```
Comment (Duhemm) [inline zinc/src/sbt-test/reporter/warnings-reported/test]: Oops, I forgot to remove that from my previous PR 😁

Comment (dwijnand) [inline zinc/src/sbt-test/reporter/warnings-reported/test]: I'll fix it up

**Comment by eed3si9n** [comment]
LGTM as is

**Comment by dwijnand** [comment]
I created a follow-on issue for the test, to (debatably) unblock this one: https://github.com/sbt/zinc/issues/154, so I'll merge.

**Pull request merged**